### PR TITLE
Fix TLS configuration for forward.

### PR DESCRIPTION
**Pull request by miekg**

### CODE REVIEW

**Comment by miekg** [comment]
Not sure if related, but:
~~~
default/coredns-deployment-664ddc7d9-htl22[coredns]: 14/Apr/2018:13:51:48 +0000 [PANIC registry-1.docker.io. AAAA] /home/miek/upstream/go/src/sync/atomic/asm_linux_arm.s:109 - runtime error: invalid memory address or nil pointer dereference
~~~

**Comment by miekg** [comment]
anyone else available for review as well? /cc @rdrozhdzh @grobie ?

File: plugin/forward/connect.go [hunk 1]
```diff
@@ -33,11 +33,15 @@ func (p *Proxy) connect(ctx context.Context, state request.Request, forceTCP, me
 	start := time.Now()
 
 	proto := state.Proto()
-	if forceTCP {
+	if p.tlsConfig != nil {
+		proto = "tcp-tls"
+	}
+
+	if forceTCP && proto == "udp" {
```
Comment (rdrozhdzh) [inline plugin/forward/connect.go:40]: this overwrites "tcp-tls" if it was set 3 lines above. Consider `else if`

File: plugin/forward/setup.go [hunk 2]
```diff
@@ -136,14 +140,25 @@ func parseForward(c *caddy.Controller) (*Forward, error) {
...
+	// Update the proxies with possibly new TLS Configs.
 	for i := range f.proxies {
-		// Only set this for proxies that need it.
-		if protocols[i] == TLS {
-			f.proxies[i].SetTLSConfig(f.tlsConfig)
+		pr := f.proxies[i]
+		pr.setExpire(f.expire)
+
+		if f.tlsConfig != nil { // forward changes
+			pr.setTLSConfig(f.tlsConfig)
```
Comment (rdrozhdzh) [inline plugin/forward/setup.go:153]: this will turn non-tls upstreams to tls, since you removed `if protocols[i] == TLS`

**Review by rdrozhdzh — CHANGES_REQUESTED** [review]
What was the reason of moving tlsConfig from transport to Proxy? Now we have to pass it in each `dial()`. IMO the transport related changes were not necessary.

**Comment by rdrozhdzh** [comment]
I think the basic fix for #1672 could be like adding
```
if f.tlsConfig == nil {
	f.tlsConfig = new(tls.Config)
}

```
before 
https://github.com/coredns/coredns/blob/9a7e487a5aac4643baf6b385cc2d8c444e648221/plugin/forward/setup.go#L138-L140

plus the change you made in Proxy.setTLSConfig() to update tlsClient

**Comment by miekg** [comment]
Thanks for the review @rdrozhdzh . You raise a good point with moving the tls from transport, that should have been necessary. Let me see if I can put that back easily.

**Comment by miekg** [comment]
Closing in favor of https://github.com/coredns/coredns/pull/1714

Thanks @rdrozhdzh, really helpful review - closing this PR.

**Pull request closed without merging**